### PR TITLE
Skip cacheline concurrency global lock in fast path

### PR DESCRIPTION
**Pull request by arutk**
The main purpose of cacheline concurrency global lock
is to eliminate the possibility of deadlocks when
locking multiple cachelines.

Cacheline lock fast path does not need to acquire
this lock, as it is only opportunistically attempting
to lock all clines without wait. There is no risk
of deadlock, as:
 * concurrent fast path will also only try_lock
   cachelines, releasing all acquired locks if failed
   to immediately acquire lock for any cacheline
 * concurrent slow path is guaranteed to have
   precedence in lock acquisition when conditions
   for deadlock occure (both slowpath and fastpath
   have acquired some locks required by the other
   thread). This is because the fastpath thread will
   back off (release acquired locks) if any one of the
   cacheline locks is not acquired.

Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>

### CODE REVIEW

**Review by imjfckm — DISMISSED** [review]

File: src/concurrency/ocf_cache_line_concurrency.c [hunk 1]
```diff
@@ -833,9 +833,7 @@ int ocf_req_async_lock_rd(struct ocf_request *req, ocf_req_async_lock_cb cb)
 		req->cache->device->concurrency.cache_line;
 	int lock;
 
-	env_rwsem_down_read(&c->lock);
 	lock = _ocf_req_trylock_rd(req);
```
Comment (arutk) [inline src/concurrency/ocf_cache_line_concurrency.c:836]: After consulting with Mariusz: looks like there is a potential for deadlock in this change when fastpath wins race with concurrent slowpath and then releases the problematic lock in fastpath rollback without transfering lock ownership to the first waiter.

**Comment by arutk** [comment]
retest this please

File: src/concurrency/ocf_cache_line_concurrency.c (hunk 1 above)
Comment (arutk) [reply src/concurrency/ocf_cache_line_concurrency.c]: if's fixed now + tests are added to validate

**Review by mariuszbarczak — APPROVED** [review]

**Review by rafalste — APPROVED** [review]

**Review by mariuszbarczak — APPROVED** [review]

**Review by rafalste — APPROVED** [review]

**Review by imjfckm — COMMENTED** [review]
🚀

**Review by imjfckm — APPROVED** [review]

**Pull request merged**